[#42] feat : 관리자 기능 추가(회원 BAN, 회원 상세조회)

kihyuk-jeong opened the pull request:
회원 BAN의 경우 BAN 상세 사유는 추후 리팩토링으로 진행하겠습니다.

1. 회원 상세 조회 
2. 회원 BAN 기능 추가

kihyuk-jeong committed:
[#42] feat : 관리자 기능 추가(회원 BAN, 회원 상세조회)

1. 회원 상세 조회 기능
2. 회원 BAN 기능
기타 : UserNotFound Exception 상태 코드 404로 변경

 #42 : Admin 기능 추가 : 사용자 BAN, 회원 상세 조회

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/AdminApiController.java:
```diff
@@ -24,4 +30,17 @@
+    @GetMapping("/users/{id}")
+    public ResponseEntity<UserDetailsResponse> getUserDetails(@PathVariable Long id) {
+        UserDetailsResponse userDetailsResponse = adminService.getUser(id);
+        return ResponseEntity.ok(userDetailsResponse);
```
여기는 `ResponseEntity`를 리턴할 필요가 없어보입니다~

kihyuk-jeong replied:
리턴타입 수정하겠습니다~

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/dto/UserDto.java:
```diff
@@ -226,4 +225,47 @@ public UserSearchCondition(Long id, String email, UserLevel userLevel) {
+        @JsonFormat
+        private LocalDateTime createDate;
+        private UserLevel userLevel;
+        private boolean isBan;
```
변수에는 `is`를 붙여주지 않아도 됩니다~ getter의 이름이 `isIsBan`이 되겠네요

f-lab-dev commented on src/main/java/com/flab/shoeauction/domain/users/user/User.java:
```diff
@@ -102,4 +102,22 @@ public User(String email, String password, UserLevel userLevel, String nickname,
+    }
+
+    public void updateBan() {
+       isBan = !isBan;
```
이런것은 명시적으로 값을 입력받도록 하는게 좋습니다~ 내부의 값이 어떤 값인지 암묵적이라 개발자의 실수가 일어날 수 있습니다

kihyuk-jeong replied:
무조건 짧게 짜는게 좋은게 아니군요 명시적으로 true / false로 처리될 수 있도록 변경하겠습니다.

kihyuk-jeong committed:
[#42] refactor : 리뷰반영

1. 변수명 수정 (isBan -> ban)
2. update 메소드 명시적으로 수정
3. 불필요한 ResponseEntity 타입 제거

 #42 : Admin 기능 추가 : 사용자 BAN, 회원 상세 조회

f-lab-dev commented on src/main/java/com/flab/shoeauction/domain/users/user/User.java:
```diff
@@ -113,11 +112,15 @@ public UserDetailsResponse toUserDetailsDto() {
 
     public void updateBan() {
-       isBan = !isBan;
+        if (ban) {
+            ban = false;
+        } else {
+            ban = true;
+        }
```
아 이런 의도가 아니였는데 잘못 전달되었나보네요~
이 메소드를 호출하는 사람은 ban의 값이 현재 무엇인지 명시적으로 알지 못합니다. 그렇기때문에 어떤 값으로 업데이트할건지는 인자로 받는게 나을 것 같다는 말이였습니다~

kihyuk-jeong replied:
제가 이해를 잘못한것 같습니다!
`UserStatus`라는 enum값을 추가하여 BAN 여부를 관리하고, BAN 요청시 request body에  BAN 시킬 ID[PK]값과 명시적인 값을 함께 전달하도록 수정하겠습니다.

kihyuk-jeong committed:
[#42] refactor : 리뷰 반영

1. BAN 여부를 UserStatus라는 enum클래스로 관리
2. BAN 요청시 명시적으로 BAN / NORMAL 값을 body에 함께 전달

kihyuk-jeong committed:
[#42] refactor : 로직 수정

1.User class에  BAN여부 검사하는 메소드 추가

f-lab-dev commented on src/main/java/com/flab/shoeauction/domain/users/common/UserStatus.java:
```diff
@@ -0,0 +1,10 @@
+ * BAN : 관리자에 의해 사용지 정지된 사용자
+ */
+public enum UserStatus {
+    NORMAL, BAN
```
혹시 ban 이외의 상태가 있을 수 있어서 enum으로 처리하신걸까요?
제가 말씀드린건 그냥 setter 형식으로 ban을 인자로 명시적으로 받아서 바꿀 수 있게 해달라는거였습니다~

kihyuk-jeong replied:
현재는 `BAN`과 `NORMAL` 뿐이지만 추후에 휴먼 계정 상태를 추가할 수도 있고, 현재는 회원 탈퇴 방식이 DELETE 쿼리를 날려 바로 삭제하는 방식인데 대부분의 사이트에서 회원 탈퇴 시 바로 삭제하지 않고 회원 상태를 변경하는 방식을 사용하고 있어서 변경 가능성이 있다고 생각하여 enum으로 처리했습니다.
